### DISCUSSION TITLE
trace_id and span_id is empty in logs auto instrument

### DISCUSSION THREAD

**Discussion by sulthans** [open-telemetry/opentelemetry-java-instrumentation#12277] [Q&A]
Gradle dependency added,

    implementation "io.opentelemetry:opentelemetry-api"
    implementation 'io.opentelemetry:opentelemetry-extension-trace-propagators'
    implementation 'io.opentelemetry:opentelemetry-sdk-extension-autoconfigure'
    implementation 'io.opentelemetry:opentelemetry-sdk'

Log4j2.xml

   ```
<Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="${text.log.pattern}" />
            <!--PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} traceId: %X{trace_id} spanId: %X{span_id} - %msg%n" /-->
        </Console>
        <OpenTelemetry name="OpenTelemetryAppender" captureMapMessageAttributes="true" captureMarkerAttribute="true" captureContextDataAttributes="*"/>
    </Appenders>

<Root level="INFO">
            <AppenderRef ref="Console"/>
            <AppenderRef ref="OpenTelemetryAppender" />
        </Root>
```

created OpenTelemetryAppender config class for Spring boot,

`@Configuration
public class OpenTelemetryConfig {

    @Bean
    OpenTelemetry openTelemetry(SdkLoggerProvider sdkLoggerProvider, SdkTracerProvider sdkTracerProvider, ContextPropagators contextPropagators) {
        OpenTelemetrySdk openTelemetrySdk = OpenTelemetrySdk.builder()
                .setLoggerProvider(sdkLoggerProvider)
                .setTracerProvider(sdkTracerProvider)
                .setPropagators(contextPropagators)
                .build();
        OpenTelemetryAppender.install(openTelemetrySdk);
        return openTelemetrySdk;
    }
}`

logs coming as,

trace_id= span_id= trace_flags=

**Comment by laurit** [comment]
The `OpenTelemetryAppender` you are using is for exporting logs via OTLP. To expose `trace_id` and `span_id` in MDC you'll need
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/log4j/log4j-context-data/log4j-context-data-2.17/library-autoconfigure Keep in mind that `trace_id` and `span_id` are only set when there is an ongoing trace. It is easier to get started using the java agent or spring boot starter https://opentelemetry.io/docs/zero-code/java/

**Reply by sulthans** [reply]
Hi @laurit Thanks for the reply, I have made the changes, still the trace_id is not added to logs and now getting below error. Trying alone for logging trace_id and span_id and add the trace_it to kafka topic header. 
Is it possible to configure to add the same trace id for kafka tracing while producing the messages to topic?

Added telemetry dependencies,

Parent : 

```
dependencySet(group: 'io.opentelemetry', version: '1.41.0') {
                entry 'opentelemetry-api'
                entry 'opentelemetry-context'
                entry 'opentelemetry-exporter-otlp'
                entry 'opentelemetry-extension-trace-propagators'
                entry 'opentelemetry-sdk'
                entry 'opentelemetry-bom'
            }

            dependencySet(group: 'io.opentelemetry', version: '1.21.0-alpha') {
                entry 'opentelemetry-semconv'
                //entry 'opentelemetry-sdk-extension-autoconfigure'
            }

            dependency group: 'io.opentelemetry.instrumentation', name: 'opentelemetry-instrumentation-annotations', version: '1.23.0'
            //dependency group: 'io.opentelemetry.instrumentation', name: 'opentelemetry-kafka-clients-2.6', version: '2.6.0-alpha'
            dependency group: 'io.opentelemetry.instrumentation', name: 'opentelemetry-log4j-appender-2.17', version: '2.6.0-alpha'
```


child : 
```
implementation 'io.opentelemetry.instrumentation:opentelemetry-instrumentation-annotations'
    implementation "io.opentelemetry:opentelemetry-api"
implementation "io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter"
    runtimeOnly "io.opentelemetry.instrumentation:opentelemetry-log4j-context-data-2.17-autoconfigure:2.6.0-alpha"
```

Do i need to run otel-collector?

trace_id= span_id= trace_flags= ERROR context.initializer.AppConfigResourceDownload - message="Could not download remote resources." FileName=appconfig_truststore.jks
org.springframework.web.client.HttpClientErrorException$NotFound: 404 Not Found: [no body]
.
.
at org.springframework.boot.SpringApplication.applyInitializers(SpringApplication.java:604) ~[spring-boot-2.7.11.jar:2.7.11]
	at org.springframework.boot.SpringApplication.prepareContext(SpringApplication.java:373) ~[spring-boot-2.7.11.jar:2.7.11]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:306) ~[spring-boot-2.7.11.jar:2.7.11]
	at com.test.Application.main(Application.java:29) ~[main/:?]


Also more error logs from stacktrace
  2024-09-18:01:13.471 [OkHttp http://localhost:4318/...] - , trace_id= span_id= trace_flags= ERROR internal.http.HttpExporter - Failed to export logs. The request could not be executed. Full error message: Failed to connect to localhost/[0:0:0:0:0:0:0:1]:4318
java.net.ConnectException: Failed to connect to localhost/[0:0:0:0:0:0:0:1]:4318
